### PR TITLE
Whats new notification

### PR DESCRIPTION
**Pull request by npretto**
Epic: https://github.com/metabase/metabase/issues/33165

Product doc: https://www.notion.so/metabase/Provide-a-heads-up-when-Metabase-is-upgraded-with-what-s-new-44a28b55624f48e28dc3f616ce7cda0f

This PR adds a notification that links to the release notes.
The notification is shown if:
- the current version is present in `version-info` (this excludes snapshots, release candidates and betas)
- there's a version in `version-info` that:
  - has a announcement_url (this will be added only to major/golden releases)
  - is > than the last acknowledged version
  - is <= the current version
  
The notification is dismissible per user and will not be dismissed automatically.
It should not cover any content in the nav bar.

## How to test locally with cypress:
This features requires the current version to be released (to be in version-info), locally this is never the case as the version name contains "SNAPSHOT".

I created a cypress test that sets up the required settings so that we can test it locally, to use it we need to:
- run metabase locally on this branch (`yarn dev`) in a terminal
- run cypress (`yarn test-cypress-open`) in another terminal
- edit the file `frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx` and replace `it.skip` with `it.only` on line 27

image

### CODE REVIEW

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    return null;
+  }
+
+  return (
+    <Paper
+      shadow="md"
+      p="md"
+      m="lg"
+      withBorder
+      style={{ borderColor: "#F0F0F0" }}
```
Comment (npretto) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx:68]: This is the color used in figma, I couldn't find it in the theme though

File: frontend/src/metabase/nav/components/WhatsNewNotification/utils.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      // shortcircuit lower bound if lastAcknowledgedVersion is null
+      (lastAcknowledgedVersion == null ||
+        //                            version > lastAcknowledgedVersion
+        MetabaseUtils.compareVersions(version, lastAcknowledgedVersion) === 1)
+    );
+  });
+
+  return (
+    eligibleVersions
+      // .sort((a, z) => z.version - a.version)
```
Comment (npretto) [inline frontend/src/metabase/nav/components/WhatsNewNotification/utils.ts:26]: we should not need this line as they're already sorted in version-info, do you think we should still sort them manually?

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        <Flex justify="space-between">
+          <Sparkles color={color("brand")} />
+          <IconButtonWrapper onClick={dimiss}>
+            <Icon name="close" color={color("bg-dark")} />
+          </IconButtonWrapper>
+        </Flex>
+
+        <Text weight="bold" color="text.2">{t`Metabase has been updated`}</Text>
+
+        <Text
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx:81]: `Anchor`

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    >
+      <Stack spacing="sm">
+        <Flex justify="space-between">
+          <Sparkles color={color("brand")} />
+          <IconButtonWrapper onClick={dimiss}>
+            <Icon name="close" color={color("bg-dark")} />
+          </IconButtonWrapper>
+        </Flex>
+
+        <Text weight="bold" color="text.2">{t`Metabase has been updated`}</Text>
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx:79]: `text.2` is the default color

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx (hunk 1 above)
Comment (ranquild) [reply frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx:69]: @npretto ask designers if they have plans to support `Paper` in our design system

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+export function WhatsNewNotification() {
+  const dispatch = useDispatch();
+  const isEmbedded = useSelector(getIsEmbedded);
+  const versionInfo = useSelector(state => getSetting(state, "version-info"));
+  const currentVersion = useSelector(state => getSetting(state, "version"));
+  const lastAcknowledgedVersion = useSelector(getLastAcknowledgedVersion);
+
+  const url: string | undefined = useMemo(() => {
+    if (isEmbedded || currentVersion.tag === undefined) {
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx:27]: Could we move all this logic to the utility function so we can test if all there?

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  createMockSettingsState,
+  createMockState,
+} from "metabase-types/store/mocks";
+import { renderWithProviders, screen } from "__support__/ui";
+import * as domUtils from "metabase/lib/dom";
+import { WhatsNewNotification } from "./WhatsNewNotification";
+
+const notification = () => screen.queryByText("See what's new");
+
+const render = ({
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx:16]: should be called `setup`

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  embedded = false,
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx:17]: we use `is` for booleans, e.g. `isEmbedded`

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  ],
+}: {
+  embedded?: boolean;
+  currentVersion?: string;
+  lastAcknowledged?: string | null;
+  versions?: VersionInfoRecord[];
+}) => {
+  jest.spyOn(domUtils, "isWithinIframe").mockReturnValue(embedded);
+
+  const versionMock = { tag: currentVersion };
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx:36]: We don't create objects for types in `metabase-types` directly. Please use the dedicated `createMock...` helper. If there is none, please create it.

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    "version-info": createMockVersionInfo({ latest, older }),
+    "last-acknowledged-version": lastAcknowledged,
+  });
+
+  return renderWithProviders(<WhatsNewNotification></WhatsNewNotification>, {
+    storeInitialState: createMockState({ settings: mockSettings }),
+  });
+};
+
+describe("WhatsNewNotification", () => {
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx:50]: Could you add more tests according to the testing plan?

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx [hunk 10]
```diff
@@ -0,0 +1,92 @@
+import { Paper } from "@mantine/core";
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx:1]: No imports from `@mantine` are allowed

File: frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
+import { createMockVersionInfoRecord } from "metabase-types/api/mocks";
+import { getLatestEligibleReleaseNotes } from "./utils";
+
+describe("What's new - utils", () => {
+  describe("getLatestEligibleReleaseNotes", () => {
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts:5]: Please add tests according to the testing plan

File: frontend/src/metabase/selectors/settings.ts [hunk 12]
```diff
@@ -107,3 +107,6 @@ export const getIsPaidPlan = createSelector(
     return tokenStatus != null && tokenStatus.valid;
   },
 );
+
+export const getLastAcknowledgedVersion = (state: State) =>
```
Comment (ranquild) [inline frontend/src/metabase/selectors/settings.ts:111]: We don't add simple selectors for settings here. If needed, add one in `nav` or in the component file

**Review by ranquild — COMMENTED** [review]
Good overall. We code move more logic to the utility function and add more tests

File: frontend/src/metabase-types/api/settings.ts [hunk 13]
```diff
@@ -108,6 +108,7 @@ export interface VersionInfoRecord {
   released?: string; // year-month-day
   patch?: boolean;
   highlights?: string[];
+  releaseNotesUrl?: string;
```
Comment (npretto) [inline frontend/src/metabase-types/api/settings.ts:111]: will field name will be changed in another PR

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx (hunk 9 above)
Comment (npretto) [reply frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx:51]: I added a few tests that were missing and now it should cover everything, do you see some test missing?

File: frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts (hunk 11 above)
Comment (npretto) [reply frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts:5]: Actually everything is/should already be tested in the component, I wrote these tests to help me write the util function without having to think about the component.
Should I remove these tests to avoid the duplication and confusion?

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+          rel="noreferrer"
+        >
+          {t`See what's new`}
+        </Anchor>
+      </Stack>
+    </NotificationContainer>
+  );
+}
+
+const NotificationContainer = styled(Box)(({ theme }) => ({
```
Comment (npretto) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx:84]: This could use a component from metabase/ui in the future: https://metaboat.slack.com/archives/C057WD5L0JG/p1692962869027779

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+  const versionInfo = useSelector(state => getSetting(state, "version-info"));
+  const currentVersion = useSelector(state => getSetting(state, "version"));
+  const lastAcknowledgedVersion = useSelector(state =>
+    getSetting(state, "last-acknowledged-version"),
+  );
+
+  const url: string | undefined = useMemo(() => {
+    const versionsList = [versionInfo?.latest]
+      .concat(versionInfo?.older)
+      .filter(isNotFalsy);
```
Comment (npretto) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx:28]: this is to make sure we're not sending undefined values while loading (and to make ts happy)

Comment (ranquild) [reply frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.tsx:28]: Could you please move the setting value and this check to the utility function?

File: frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts [hunk 16]
```diff
@@ -0,0 +1,102 @@
+import { createMockVersionInfoRecord } from "metabase-types/api/mocks";
+import { getLatestEligibleReleaseNotes } from "./utils";
+
+describe("What's new - utils", () => {
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts:4]: nit: we don't add an enclosing `describe` block and start with `describe` for what's being tested directly

File: frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+              announcement_url: "url",
+            }),
+          ],
+          currentVersion: "v0.48",
+          lastAcknowledgedVersion: "v0.47",
+        }),
+      ).not.toBe(undefined);
+    });
+
+    it("filters out future versions", () => {
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts:65]: This test doesn't work as expected. The version record is filtered out because there is no version record for v0.47

File: frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts (hunk 11 above)
Comment (ranquild) [reply frontend/src/metabase/nav/components/WhatsNewNotification/utils.unit.spec.ts:5]: There are no tests for 
1 - for embedding nothing is returned even if all conditions are otherwise correct
2 - if the current version is not defined
3 - if there is no version record for the current version available

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+  describe("display logic", () => {
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx:53]: Most of this tests could be moved to the utility function because there is no code in the component that does this

File: frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+      setup({
+        isEmbedded: true,
+        currentVersion: "v0.48.0",
+        lastAcknowledged: null,
+      });
+      expect(notification()).not.toBeInTheDocument();
+    });
+  });
+
+  describe("link behaviour", () => {
```
Comment (ranquild) [inline frontend/src/metabase/nav/components/WhatsNewNotification/WhatsNewNotification.unit.spec.tsx:126]: There is no test that by clicking on the dismiss button the BE endpoint is called

File: e2e/test/scenarios/onboarding/navbar/whats-new.cy.spec.js [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+      res.body["version-info"] = createMockVersionInfo({ latest, older });
+    });
+  }).as("sessionProperties");
+}
+
+function loadHomepage() {
+  cy.visit("/");
+
+  // make sure page is loaded
+  cy.findByText("loading").should("not.exist");
```
Comment (ranquild) [inline e2e/test/scenarios/onboarding/navbar/whats-new.cy.spec.js:80]: These should probably go in reverse order - with `wait` first. The difference is that `cy.wait` waits for much longer is the request was made. We have many flaky tests because of regular `findByText` checks timeouts

File: e2e/test/scenarios/onboarding/navbar/whats-new.cy.spec.js [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+        mockVersion({
+          version: "v0.47.1",
+          announcement_url: "metabase.com/releases/47",
+        }),
+      ],
+    });
+  });
+
+  // replace `skip` with `only` to manually test this feature with the mocks set up
+  it.skip("mock setup for manual tests", () => {
```
Comment (ranquild) [inline e2e/test/scenarios/onboarding/navbar/whats-new.cy.spec.js:27]: Please remove this before merging

**Review by ranquild — APPROVED** [review]
Please take a look at what's being tested and where. In most cases we should add tests covering logic present in the tested function/component

File: e2e/test/scenarios/onboarding/navbar/whats-new.cy.spec.js (hunk 21 above)
Comment (npretto) [reply e2e/test/scenarios/onboarding/navbar/whats-new.cy.spec.js:27]: Should I remove it before merging in the feature branch or before merging in master?
